File: 0376-wiggle-subsequence/0376-wiggle-subsequence.py
class Solution(object):
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        n = len(nums)
        # 处理trivial 情况
        if n == 1:
            return 1
        
        if n == 2:
            return 2 if nums[0] != nums[1] else 1
        
        # 删除相邻的重复元素
        new_nums = [nums[0]]
        for i in range(1, n):
            if nums[i] != nums[i - 1]:
                new_nums.append(nums[i])
        
        if len(new_nums) == 1:
            return 1
        
        res = 2
        for i in range(1, len(new_nums) - 1):
            a = new_nums[i - 1]
            b = new_nums[i]
            c = new_nums[i + 1]
            if (b > a and b > c) or (b < a and b < c):
                res += 1
                
        return res
            
        
        
        

File: 0376-wiggle-subsequence/0376-wiggle-subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        nums.erase(unique(nums.begin(), nums.end()), nums.end());
        if (nums.size() <= 2) return nums.size();
        int res = 2;
        for (int i = 1; i + 1 < nums.size(); i ++ ) {
            int a = nums[i - 1], b = nums[i], c = nums[i + 1];
            if (b > a && b > c || b < a && b < c) res ++ ;
        }
        return res;
    }
};

